### ISSUE TITLE
How does concurrency control work?

### ISSUE THREAD

**Issue by zergon321** [ClickHouse/ClickHouse#13109]
Hi. I'd like to know how **ClickHouse** ensures *concurrency control* at table level. For example, if there's many applications writing in the same table at the same time, won't data get corrupted?

**Comment by alexey-milovidov** [comment]
MergeTree tables are represented by a set of immutable data parts. Data parts are never modified, only created and deleted. INSERTs create new data parts. SELECTs take snapshot of required data parts and hold the snapshot during the query.
Snapshots are based on reference counting and snapshot acquire operation is very cheap.
Both SELECTs and INSERTs and even ALTERs can be executed concurrently.

Data parts are merged in background - a new larger data part is created as a result of merge, then source data parts are deleted.
When data needs to be modified (ALTER query), we perform copy-on-write - create another part and remove source part.
Version number and block number are tracked for each part.

TLDR, it's MVCC.

**Comment by zergon321** [comment]
Thank you very much for the answer!
